### PR TITLE
Feat: Validate input csv file for invalid emails

### PR DESCRIPTION
**Pull request by codemage8**
https://www.loom.com/share/4d149751c4a54dda8b585301f01308cc

@tianhuil I would like to hear your ux-feedback about invalid emails

### CODE REVIEW

File: lib/parse.ts [hunk 1]
```diff
@@ -2,10 +2,22 @@ import papaparse from 'papaparse'
 import type { UserRow, UsersData } from '~/lib/hooks'
 import { atOrThrow } from '~/lib/util'
 
+const emailRegEx = /^([a-zA-Z0-9._%-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,})$/
+
+export class InvalidEmailsError extends Error {
+  constructor(public readonly invalidRows: [number, UserRow][]) {
+    if (!invalidRows.length) {
+      throw new Error('No errors')
+    }
```
Comment (tianhuil) [inline lib/parse.ts]: If there are no errors, you should probably not throw an error.

File: lib/parse.ts [hunk 2]
```diff
@@ -25,6 +37,15 @@ export const parseFile = (file: File | string): Promise<UsersData> =>
           reject(new Error('No user rows found'))
           return
         }
+        const invalidRows = rows.reduce<[number, UserRow][]>((current, row, index) => {
+          if (!row.email.match(emailRegEx)) {
+            return [...current, [index, row]]
+          }
+          return current
+        }, [])
```
Comment (tianhuil) [inline lib/parse.ts]: To make this easier to read, let's make it `filter` and then `map`, not `reduce`.

**Review by tianhuil — APPROVED** [review]
Some comments.

File: lib/parse.ts (hunk 2 above)
Comment (codemage8) [reply lib/parse.ts]: > To make this easier to read, let's make it `filter` and then `map`, not `reduce`.

If we use `filter` at first, then it's a bit tricky to save the invalid row indexes

Comment (tianhuil) [reply lib/parse.ts]: Does this work: https://github.com/zilinj88/nextjs-mailmerge/pull/14/commits/d4b31189690e86e0973575455e658a61aa552049?

**Pull request merged**